section_faiss_vector_index.go

```diff
@@ -382,15 +388,42 @@ func (v *vectorIndexOpaque) grabBuf(size int) []byte {
 	return buf[0:size]
 }
 
-func getIndexType(nVecs int) (string, bool) {
+// Return the number of centroids for an IVF index.
+func getNumCentroids(nVecs int) int {
+	var nlist int
+
+	switch {
+	// At 1M vectors, nlist = 4k gave a reasonably high recall with the right nprobe,
+	// whereas 1M/100 = 10000 centroids would increase training time without
+	// corresponding increase in recall
+	case nVecs >= 1000000:
+		nlist = int(4 * math.Sqrt(float64(nVecs)))
+	case nVecs >= 1000:
+		// 100 points per cluster is a reasonable default, considering the default
+		// minimum and maximum points per cluster is 39 and 256 respectively.
+		// Since it's a recommendation to have a minimum of 10 clusters, 1000(100 * 10)
+		// was chosen as the lower threshold.
+		nlist = nVecs / 100
+	}
+	return nlist
+}
+
+const (
+	IndexTypeFlat = iota
+	IndexTypeIVF
+)
+
+// Returns a description string for the index and quantizer type
+// and an index type.
+func getIndexType(nVecs, nlist int) (string, int) {
 	switch {
-	case nVecs > 10000:
-		return "IVF100,SQ8", true
-	// default minimum number of training points per cluster is 39
-	case nVecs >= 390:
-		return "IVF10,Flat", true
+	case nVecs >= 10000:
+		return "IVF" + strconv.Itoa(nlist) + ",SQ8", IndexTypeIVF
```

Comment:
Better move would be to

```
return fmt.Sprintf("IVF%d,SQ8", nlist)
``` 

.. rather than declaring 3 strings and concats. Same below.

Reply:
yeah, my bad, good catch.